ocaml/gencert/lib.ml

```diff
@@ -65,107 +65,36 @@ let get_management_ip_addr ~dbg =
 
 open Api_errors
 open Rresult
+open Validation
 
-type t_certificate = Leaf | Chain
-
-let validate_private_key pkcs8_private_key =
-  let ensure_key_length = function
-    | `RSA priv ->
-        let length = Mirage_crypto_pk.Rsa.priv_bits priv in
-        if length < 2048 || length > 4096 then
-          Error
-            (`Msg
-              ( server_certificate_key_rsa_length_not_supported
-              , [Int.to_string length] ))
-        else
-          Ok (`RSA priv)
-  in
-  let raw_pem = Cstruct.of_string pkcs8_private_key in
-  X509.Private_key.decode_pem raw_pem
-  |> R.reword_error (fun (`Msg err_msg) ->
-         let unknown_algorithm = "Unknown algorithm " in
-         if Astring.String.is_prefix ~affix:"multi-prime RSA" err_msg then
-           `Msg (server_certificate_key_rsa_multi_not_supported, [])
-         else if Astring.String.is_prefix ~affix:unknown_algorithm err_msg then
-           `Msg
-             ( server_certificate_key_algorithm_not_supported
-             , [
-                 Astring.String.with_range
-                   ~first:(String.length unknown_algorithm)
-                   err_msg
-               ] )
-         else (
-           D.info {|Failed to validate private key because "%s"|} err_msg ;
-           `Msg (server_certificate_key_invalid, [])))
-  >>= ensure_key_length
-
-let validate_certificate kind pem now private_key =
-  let ensure_keys_match private_key certificate =
-    let public_key = X509.Certificate.public_key certificate in
-    match (public_key, private_key) with
-    | `RSA pub, `RSA priv when pub = Mirage_crypto_pk.Rsa.pub_of_priv priv ->
-        Ok certificate
-    | _ ->
-        Error (`Msg (server_certificate_key_mismatch, []))
+let install_server_certificate ?(pem_chain = None) ~pem_leaf ~pkcs8_private_key
```

Comment:
The validations imply that an invalid cert cannot be installed. There is some friction with the startup behaviour where we also verify the certificate but in addition check if TLS verification is on and currently would install the certificate if TLS is off. I believe this needs to be simplified and would suggest to never install a certificate that is known to be invalid even with TLS validation off as it would create problems down the road.

Reply:
I guess on startup if tls verification is off nothing is done; or else the pem is read and installed using `install_server_certificate`. It adds some unnecessary writing, but it adds more regularity. The exceptions have to be caught correctly so they don't interrupt the startup process, however.

Reply:
The current plan is the check the validity of the PEM at startup. If the verification code in this function could be pulled out, it could be re-used. Otherwise it would have to be replicated as this function verifies the certificate and installs it.

Reply:
I think exposing the chain validation functionality by itself is perfectly acceptable, that should be enough to validate the PKCS12 file present in the filesystem at startup.

Reply:
I've offloaded all the decoding and validation to the function `validating_trust_anchors`. This returns the decoded server certificate and the list of validating anchors, which is allowed to be empty.

I'm pondering whether this should be allowed, as it means the certificate could not be validated and therefore it should return an error instead.